pkg/capacity/topology/nodes_test.go

```diff
@@ -0,0 +1,535 @@
+			initialNodes: []testNode{
+				{
+					name: node1,
+					driverKeys: map[string][]string{
+						driverName: localStorageKeys,
+					},
+					labels: localStorageLabelsNode1,
+				},
+			},
+			expectedSegments: []*Segment{localStorageNode1},
+			update: func(t *testing.T, client *fakeclientset.Clientset) {
+				removeNode(t, client, node1)
+			},
+		},
+		"remove-driver": {
+			initialNodes: []testNode{
+				{
+					name: node1,
+					driverKeys: map[string][]string{
+						driverName: localStorageKeys,
+					},
+					labels: localStorageLabelsNode1,
+				},
+			},
+			expectedSegments: []*Segment{localStorageNode1},
+			update: func(t *testing.T, client *fakeclientset.Clientset) {
+				csiNode, err := client.StorageV1().CSINodes().Get(context.Background(), node1, metav1.GetOptions{})
+				if err != nil {
+					t.Fatalf("unexpected error: %v", err)
+				}
+				csiNode.Spec.Drivers = nil
+				if _, err := client.StorageV1().CSINodes().Update(context.Background(), csiNode, metav1.UpdateOptions{}); err != nil {
+					t.Fatalf("unexpected error: %v", err)
+				}
+			},
+		},
+		"change-labels": {
+			initialNodes: []testNode{
+				{
+					name: node1,
+					driverKeys: map[string][]string{
+						driverName: localStorageKeys,
+					},
+					labels: localStorageLabelsNode1,
+				},
+			},
+			expectedSegments: []*Segment{localStorageNode1},
+			update: func(t *testing.T, client *fakeclientset.Clientset) {
+				node, err := client.CoreV1().Nodes().Get(context.Background(), node1, metav1.GetOptions{})
+				if err != nil {
+					t.Fatalf("unexpected error: %v", err)
+				}
+				// This isn't a realistic test case because CSI drivers cannot change their topology?
+				// We support it anyway.
+				node.Labels[localStorageKey] = node2
+				if _, err := client.CoreV1().Nodes().Update(context.Background(), node, metav1.UpdateOptions{}); err != nil {
+					t.Fatalf("unexpected error: %v", err)
+				}
+			},
+			expectedUpdatedSegments: []*Segment{localStorageNode2},
+		},
+	}
+
+	for name, tc := range testcases {
+		// Not run in parallel. That doesn't work well in combination with global logging.
+		t.Run(name, func(t *testing.T) {
+			// There is no good way to shut down the controller. It spawns
+			// various goroutines and some of them (in particular shared informer)
+			// become very unhappy ("close on close channel") when using a context
+			// that gets cancelled. Therefore we just keep everything running.
+			ctx := context.Background()
+			// ctx, cancel := context.WithCancel(context.Background())
+			// defer cancel()
+
+			testDriverName := tc.driverName
+			if testDriverName == "" {
+				testDriverName = driverName
+			}
+
+			var objects []runtime.Object
+			objects = append(objects, makeNodes(tc.initialNodes)...)
+			clientSet := fakeclientset.NewSimpleClientset(objects...)
+			nt := fakeNodeTopology(ctx, testDriverName, clientSet)
+
+			added, removed, _ := addTestCallback(nt)
+			go nt.Run(ctx)
+
+			time.Sleep(1 * time.Second)
```

Comment:
Me neither, but I wasn't sure how to avoid it because the informers run in goroutines. `HasSynced()` might work, let me try that.

> In addition, tc.update needs to update the informer caches too... Or should the tc include initialSegments to fill the nodeTopology and no update call is then needed?

That's where I currently need the delay: because the update method just uses the client API, the informers need to time to catch up. Updating the informers directly may be a way to avoid that delay.